src/PageParser.php

```diff
@@ -418,15 +418,9 @@ private function cleanRedLinks() {
 	}
 
 	private function cleanReferenceLinks() {
-		// Get all links that contain the "style" value "mw-Ref"
-		$links = $this->xPath->query( '//html:a[contains(@style, "mw-Ref")]' );
-		foreach ( $links as $link ) {
-			$href = $link->getAttribute( 'href' );
-			$pos = strpos( $href, '#' );
-			$link->setAttribute( 'href', substr( $href, $pos ) );
-		}
-		// Get all links that have the "rel" attribute equals to "mw-Ref"
-		$links = $this->xPath->query( '//html:a[@rel="mw:referencedBy"]' );
+		$links = $this->xPath->query(
+			'//a[contains(@style, "mw-Ref") or @rel="mw:referencedBy" or @rel="mw:WikiLink"]'
```

Comment:
As it turns out, links for some of the examples in the task did not match `a[contains(@style, "mw-Ref") or @rel="mw:referencedBy"` and instead they had  `@rel="mw:WikiLink"`

Does anyone knows why they are different depending on the source? I'm afraid we could be missing other variations

Reply:
I bet there are things we're missing. :-) I wouldn't worry about catching absolutely everything though; we can follow up with more.

Are they actual reference links? Or just normal links? Or normal links with a reference? (Which page are these on?)

Reply:
The ones in here https://wsexport.wmflabs.org/?lang=en&page=The_Variation_of_Animals_and_Plants_under_Domestication%2FXIV&format=epub-3&fonts=

I'm not sure I know the difference between a `reference link` and `normal links with a reference`. I guess it is either of those?

Reply:
Reference links are those built in Ws with ref tags to create footnotes. I believe they were catched with the cleanReferenceLinks() function.

I think that normal links are catched with the setlinks() function but it doesn't work anymore when the full path is not provided -> in this situation :  if ( substr( $href, 0, 1 ) === '#' )  ; but it does work for all the  elseif situations.

Some references are hand-built and may appear as normal links.

Reply:
I think the xpath to select all reference links and backlinks should be something like:

    //*[@typeof="mw:Extension/ref"]/a | //a[@rel="mw:referencedBy"]

Reply:
Thank you @DenisGagne52.

@samwilson that query works in most cases that I've tested but this one sleeps thru 
`<sup id="ref_136" class="plainlinks" about="#mwt7" typeof="mw:Transclusion"><a rel="mw:WikiLink" href="#endnote_136">[136]</a></sup>` taken from https://wsexport.wmflabs.org/?lang=en&page=The_Variation_of_Animals_and_Plants_under_Domestication%2FXIV&format=epub-3&fonts=

I've updated the patch accordingly

Reply:
That page is constructing 'manual' footnote links, with e.g. `[[#endnote_123|[123]]]` (and matching backlinks from the footnotes, constructed in a similar manner). This means that there's nothing particular about the links that we can match on — the `<sup typeof="mw:Transclusion">` is just a coincidence because it's in a template.

I think what we need to do is strip the page title, which is not currently available in the PageParser class. We'll have to pass it in everywhere.

A link such as `[[#Foo]]` on page `Lorem` (it doesn't matter if it has a leading pagename or alternate display text) is rendered as `<a href="./Lorem#Foo">#Foo</a>`, so we need to look for `./Lorem#` and remove everything before `#` (but not if it's e.g. `[[Lorem/Ipsum#Foo]]` because that's a different page).

I set up some tests at https://en.wikisource.beta.wmflabs.org/wiki/Links